### PR TITLE
feature/test_source_from_api

### LINKED ISSUE #3555: Add endpoint POST config/source/:id/test

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Fixes #3555

Description
===========

Add endpoint POST config/source/:id/test

### CODE REVIEW

File: lib/pf/UnifiedApi/Controller/Config/Sources.pm [hunk 1]
```diff
@@ -94,6 +94,47 @@ sub type_lookup {
     return \%TYPES_TO_FORMS;
 }
 
+sub test {
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Config/Sources.pm:103]: perldoc

File: lib/pf/UnifiedApi/Controller/Config/Sources.pm [hunk 2]
```diff
@@ -94,6 +94,47 @@ sub type_lookup {
...
+    my $success = eval {
+        my $source = newAuthenticationSource($new_data->{type}, 'test', $form->getSourceArgs());
+        my $method = $source->can('test');
+        if (!$method) {
+            return $self->render_error(HTTP_METHOD_NOT_ALLOWED, "$new_data->{type} cannot be tested");
+        }
+
+        my ($status, $message) = $source->test();
+        if (!$status) {
+            return $self->render_error(HTTP_BAD_REQUEST, $message);
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Config/Sources.pm]: Not sure 400 is the right code for a failure to test an auth source

Perhaps 422 would be better

File: lib/pf/UnifiedApi/Controller/Config/Sources.pm [hunk 3]
```diff
@@ -94,6 +94,47 @@ sub type_lookup {
...
+        }
+        return 1;
+    };
+    if ($@) {
+        return $self->render_error(HTTP_BAD_REQUEST, "$@");
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Config/Sources.pm]: Same comment about return code

Also, some places you use constants, some the number, I don't care if its any of them but I believe it should be consistent

**Comment by jrouzierinverse** [comment]
Ready for review

**Comment by julsemaan** [comment]
Does the spec need to be manually adjusted for this or will it be automagically adjusted ?

**Comment by julsemaan** [comment]
Also this needs a rebase

**Comment by jrouzierinverse** [comment]
I am confused which spec and why?

**Comment by julsemaan** [comment]
openapi spec sorry

**Comment by julsemaan** [comment]
Still missing the doc in the openapi spec

**Comment by julsemaan** [comment]
@jrouzierinverse 

Do you want to write the spec so we can merge this or should this be postponed to 8.3 ?

**Comment by extrafu** [comment]
Bump - what's the status on this one?

**Comment by jrouzierinverse** [comment]
Add API Spec for sources/test

**Comment by jrouzierinverse** [comment]
Done

**Comment by extrafu** [comment]
Bump.

File: html/pfappserver/lib/pfappserver/Form/Config/Source/Htpasswd.pm [hunk 4]
```diff
@@ -36,8 +36,8 @@ sub validate {
     my $self = shift;
 
     $self->SUPER::validate();
-
-    unless (-r $self->value->{path}) {
+    my $path = $self->value->{path};
+    unless (defined $path && -r $path) {
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/Source/Htpasswd.pm]: We should use the same styling as everywhere else when there are 2 conditions
```suggestion
    unless (defined($path) && -r $path) {
```

**Review by julsemaan — CHANGES_REQUESTED** [review]

**Comment by jrouzierinverse** [comment]
Done

**Comment by extrafu** [comment]
@julsemaan Ready for merging?

**Comment by extrafu** [comment]
@julsemaan ping!

**Pull request merged**